### PR TITLE
Homework13mb

### PR DESCRIPTION
**Pull request by Borisko130**

### CODE REVIEW

File: src/main/java/core/basesyntax/Car.java [hunk 1]
```diff
@@ -1,13 +1,84 @@
...
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = new ArrayList<>();
+        for (Wheel wheel : wheels) {
+            this.wheels.add(new Wheel(wheel.getRadius()));
+        }
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Car.java:22]: let's move this logic to separate method

File: src/main/java/core/basesyntax/Car.java [hunk 2]
```diff
@@ -1,13 +1,84 @@
...
+        if (engine != null) {
+            this.engine = new Engine(engine.getHorsePower(), engine.getMaker());
+        } else {
+            this.engine = null;
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Car.java:26]: same here

File: src/main/java/core/basesyntax/Car.java [hunk 3]
```diff
@@ -1,13 +1,84 @@
...
+        }
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(getYear(), getColor(), getWheels(),
+                new Engine(engine.getHorsePower(), engine.getMaker()));
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Car.java:32]: it' better to use engine clone here

File: src/main/java/core/basesyntax/Car.java [hunk 4]
```diff
@@ -1,13 +1,84 @@
...
+
+    public Car changeColor(String newColor) {
+        return new Car(getYear(), newColor, getWheels(), getEngine());
+
+    }
+
+    public Car addWheel(Wheel newWheel) {
+        List<Wheel> newWheelsList = getWheels();
+        newWheelsList.add(newWheel);
+        return new Car(getYear(), getColor(), new ArrayList<>(newWheelsList), getEngine());
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Car.java:43]:
```suggestion
        return new Car(getYear(), getColor(), newWheelsList, getEngine());
```

File: src/main/java/core/basesyntax/Car.java [hunk 5]
```diff
@@ -1,13 +1,84 @@
...
+
+    public List<Wheel> getWheels() {
+        return new ArrayList<>(wheels);
+    }
+
+    public Engine getEngine() {
+        if (engine != null) {
+            return engine.clone();
+        }
+        return null;
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Car.java:62]:
```suggestion
        return engine == null ? null : engine.clone();
```

File: src/main/java/core/basesyntax/Car.java [hunk 6]
```diff
@@ -1,13 +1,88 @@
...
+            this.wheels.add(new Wheel(wheel.getRadius()));
+        }
+    }
+
+    private Engine initEngine(Engine engine) {
+        if (engine != null) {
+            return new Engine(engine.getHorsePower(), engine.getMaker());
+        } else {
+            return null;
+        }
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Car.java:66]: use here ternary operator as well

File: src/main/java/core/basesyntax/Car.java [hunk 7]
```diff
@@ -1,13 +1,88 @@
...
+
+    private void initWheels(List<Wheel> wheels) {
+        for (Wheel wheel : wheels) {
+            this.wheels.add(new Wheel(wheel.getRadius()));
+        }
+    }
+
+    private Engine initEngine(Engine engine) {
+        if (engine != null) {
+            return new Engine(engine.getHorsePower(), engine.getMaker());
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Car.java:63]: use method clone here

File: src/main/java/core/basesyntax/Car.java [hunk 8]
```diff
@@ -1,13 +1,88 @@
...
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Car.java:70]: don't mix private and public methods in your class

File: src/main/java/core/basesyntax/Wheel.java [hunk 9]
```diff
@@ -1,5 +1,41 @@
...
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        Wheel wheel = (Wheel) o;
+        return getRadius() == wheel.getRadius();
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Wheel.java:29]: let's replace getters in `equals()` and `hashCode()` in all three classes

File: src/main/java/core/basesyntax/Car.java [hunk 10]
```diff
@@ -1,13 +1,84 @@
...
+        initWheels(wheels);
+        this.engine = initEngine(engine);
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(getYear(), getColor(), getWheels(), engine.clone());
+    }
+
+    public Car changeColor(String newColor) {
+        return new Car(getYear(), newColor, getWheels(), getEngine());
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:29]: and not only in `equals()` and `hashCode()`, but everywhere

File: src/main/java/core/basesyntax/Car.java [hunk 11]
```diff
@@ -1,13 +1,84 @@
...
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = new ArrayList<>();
+        initWheels(wheels);
+        this.engine = initEngine(engine);
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(getYear(), getColor(), getWheels(), engine.clone());
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:25]: we don't really need to pass the clone - in the constructor we will make a copy of engine anyway

File: src/main/java/core/basesyntax/Car.java [hunk 12]
```diff
@@ -1,13 +1,84 @@
...
+        return Objects.hash(getYear(), getColor(), getWheels(), getEngine());
+    }
+
+    private void initWheels(List<Wheel> wheels) {
+        for (Wheel wheel : wheels) {
+            this.wheels.add(new Wheel(wheel.getRadius()));
+        }
+    }
+
+    private Engine initEngine(Engine engine) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:81]:
```suggestion
    private Engine getEngineCopy(Engine engine) {
```

File: src/main/java/core/basesyntax/Car.java [hunk 13]
```diff
@@ -1,13 +1,84 @@
...
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return new ArrayList<>(wheels);
+    }
+
+    public Engine getEngine() {
+        return engine == null ? null : engine.clone();
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:52]: you already have realized the same logic in another method
let's use it

File: src/main/java/core/basesyntax/Car.java [hunk 14]
```diff
@@ -1,13 +1,84 @@
...
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = new ArrayList<>();
+        initWheels(wheels);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:20]:
```suggestion
        this.wheels = getWheelsCopy(wheels);
```
let's create `getWheelsCopy(wheels)` method that we will be able to reuse
this method should create a deep list copy - that is, create a new list and fill it with copies of all wheels

File: src/main/java/core/basesyntax/Car.java [hunk 15]
```diff
@@ -1,13 +1,84 @@
...
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return new ArrayList<>(wheels);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:48]: `new ArrayList<>(wheels)` does not create a deep copy
the list is different, but it is filled with the same Wheel objects

create `getWheelsCopy(wheels)` and use it here

**Pull request closed without merging**